Plugin: require minimum WordPress 6.2

ramonjd opened the pull request:
## What?

WordPress 6.3 has been released: https://wordpress.org/news/2023/08/lionel/ 🎹 

This PR removes 6.2 compatibility code, thereby making WordPress 6.2 the minimum version required for the Gutenberg plugin.

This is a first pass. There might be 6.2 code and other compat work to be done after this, for example, ensuring that any deprecated functions are replaced, duplicated unit tests removed.

## How?
- Removing the `compat/6.2` folder
- Replacing 6.2 `gutenberg_` methods with their core equivalents
- Removed some tests already covered in core since 6.2

## Testing Instructions
Ensure things are not broken.
Run the Gutenberg plugin with WP 6.2, and confirm editors work as before.
Run the Gutenberg plugin with WP 6.3, and confirm editors work as before.

If you're using `wp-env` you can add an `.wp-env.override.json` file to Gutenberg root:

```json
{
	"core": "WordPress/WordPress#6.2"
}
```

or

```json
{
	"core": "WordPress/WordPress#6.3"
}
```

and run `npm run wp-env start --update`

ramonjd committed:
Initial commit

ramonjd committed:
Swapping `gutenberg_` functions for wp equivalents:
- wp_get_elements_class_name since 6.0.0
- _wp_get_presets_class_name since 6.2.0

Ensuring `build_query_vars_from_query_block` doesn't receive `gutenberg_` prefix during build

ramonjd commented on lib/block-supports/elements.php:
```diff
@@ -5,16 +5,6 @@
- * @param array $block Block object.
- * @return string      The unique class name.
- */
-function gutenberg_get_elements_class_name( $block ) {
```
since 6.0.0

ramonjd commented on lib/block-supports/settings.php:
```diff
@@ -5,18 +5,6 @@
- * @param array $block Block object.
- * @return string      The unique class name.
- */
-function _gutenberg_get_presets_class_name( $block ) {
```
since 6.2.0

ramonjd committed:
- Remove compat/6.2 folder completely
- Add usages of wp_normalize_remote_block_pattern (since 6.2)

ramonjd commented on lib/compat/wordpress-6.3/block-patterns.php:
```diff
@@ -110,7 +110,7 @@ function gutenberg_load_remote_featured_patterns() {
 	foreach ( $patterns as $pattern ) {
 		$pattern['source']  = 'pattern-directory/featured'; // Added in 6.3.0.
-		$normalized_pattern = gutenberg_normalize_remote_pattern( $pattern );
+		$normalized_pattern = wp_normalize_remote_block_pattern( $pattern );
```
since 6.2

ramonjd committed:
- Using get_template_hierarchy (since 6.2)
- Removed tests already covered in Core

ramonjd commented:
Adding a few folks who worked in `compat/6.2` as reviewers

ramonjd commented:
Requires https://github.com/WordPress/gutenberg/pull/53499 for the static analysis tests to pass

dmsnell reviewed (commented):
Thanks @ramonjd! Things look okay from the perspective of the HTML API ✅

aaronrobertshaw reviewed (commented):
Nice work @ramonjd 👍 

Changes LGTM. 

I didn't spot any issues comparing patterns behaviour between trunk and this PR using both 6.2 and 6.3.

oandregal commented on lib/compat/wordpress-6.2/get-global-styles-and-settings.php:
```diff
@@ -1,87 +0,0 @@
- * @package gutenberg
- */
-
-if ( ! function_exists( 'wp_theme_has_theme_json' ) ) {
```
Removing this file is ok. It was introduced in 6.2, and it was already using the core functions, when available.

oandregal commented on lib/compat/wordpress-6.2/default-filters.php:
```diff
@@ -1,18 +0,0 @@
- * (admin area, multisite environment…), please move it
- * to a more appropriate file instead.
- *
- * @package gutenberg
```
LOL, this was empty 😄

ramonjd replied:
👻

oandregal commented on lib/compat/wordpress-6.2/block-editor.php:
```diff
@@ -1,45 +0,0 @@
- * @package gutenberg
- */
-
-if ( ! function_exists( 'get_classic_theme_supports_block_editor_settings' ) ) {
```
Sounds good.

oandregal commented on lib/compat/wordpress-6.2/class-gutenberg-rest-global-styles-controller-6-2.php:
```diff
@@ -1,304 +0,0 @@
-/**
- * Base Global Styles REST API Controller.
- */
-class Gutenberg_REST_Global_Styles_Controller_6_2 extends WP_REST_Global_Styles_Controller {
```
Instead of removing it, we need to make the global styles' endpoint evergreen, like `global-styles-and-settings.php` is. I thought it was already the case.

This is why: as long as the underlying data and how it is processed changes (`theme.json`, `WP_Theme_JSON_Gutenberg`, and `WP_Theme_JSON_Resolver_Gutenberg`), everything that sits on top of it needs to use the Gutenberg classes.

Example: imagine we add a new key `myNewKey` for `theme.json` in Gutenberg. If the REST API endpoint is not using the Gutenberg classes (schema, sanitizer, etc.) that key will be removed by the core classes. Same if we change how styles are generated, for example.

Sorry that I can pinpoint specifics of how this would break, I have a limited view of what changed as I wasn't involved in this area for the past few cycles. Hoping this helps anyway?

ramonjd replied:
Ah yes, good point, thanks. I'll fix that up.

ramonjd replied:
> Instead of removing it, we need to make the global styles' endpoint evergreen, like global-styles-and-settings.php is. I thought it was already the case.

Just looking at this now. I might do this in a separate PR so we can decide where to put it. I'm thinking `rest-api/` or something.

Making it evergreen might cause a few headaches during WP release (picking through changes) but it will ensure we don't miss stuff like this when removing compat files.

And also that we're serving up the latest changes from `_Gutenberg` theme JSON classes 😄

ramonjd replied:
I've added a PR here: https://github.com/WordPress/gutenberg/pull/53618

ramonjd committed:
Moving get_items() method from Gutenberg_REST_Block_Patterns_Controller_6_2 to Gutenberg_REST_Block_Patterns_Controller_6_3 because the following methods were updated in 6.3:

- gutenberg_load_remote_block_patterns
- gutenberg_load_remote_featured_patterns
- gutenberg_register_remote_theme_patterns

ramonjd commented on lib/compat/wordpress-6.3/class-gutenberg-rest-block-patterns-controller-6-3.php:
```diff
@@ -158,4 +158,33 @@ public function get_item_schema() {
+			gutenberg_load_remote_block_patterns(); // Patterns with the `core` keyword.
+			gutenberg_load_remote_featured_patterns(); // Patterns in the `featured` category.
+			gutenberg_register_remote_theme_patterns(); // Patterns requested by current theme.
+
```
I moved the `get_items()` method from `Gutenberg_REST_Block_Patterns_Controller_6_2` to `Gutenberg_REST_Block_Patterns_Controller_6_3` because the following methods were updated in 6.3:

- gutenberg_load_remote_block_patterns
- gutenberg_load_remote_featured_patterns
- gutenberg_register_remote_theme_patterns

The consequence is that Gutenberg plugin users who are running WordPress 6.2 will get the benefit of these 6.3 changes.

Does that sound legit @aaronrobertshaw ?

ntsekouras replied:
That's sounds right, yes.

ramonjd replied:
Thanks for checking 🙇🏻

ramonjd committed:
Referencing private property

ramonjd committed:
Delete lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-controller-6-3.php because it wanted to live after a less than ideal rebase

gziolo commented:
Are there any blockers? It would be great to land it now after RC1 for the plugin gets created, as we will have 3 weeks to catch any regressions before these changes are released.

ramonjd commented:
> Are there any blockers? It would be great to land it now after RC1 for the plugin gets created, as we will have 3 weeks to catch any regressions before these changes are released.

Thanks for the ping @gziolo 

None that I'm aware of, just looking for a ✅ 

I'll rebase today and give it a final smoke test.

ramonjd commented:
PHP 7 fails unrelated. Suspect recent Core changes: https://core.trac.wordpress.org/changeset/56500

gziolo reviewed (approved):
We definitely need to wait until all CI checks pass, but otherwise, this PR seems ready. It looks similar to all other changes applied when increasing the minimum WP version supported.

ntsekouras reviewed (approved):
Smoke tested and looks good! Let's 🚢 . Thank you Ramon!
